src/drivers/intel/cavs/dmic.c

```diff
@@ -1583,6 +1584,97 @@ static int dmic_get_fifo(struct dai *dai, int direction, int stream_id)
 	return dai->plat_data.fifo[SOF_IPC_STREAM_CAPTURE].offset;
 }
 
+/* Functions for HW timestamp */
+
+static inline uint32_t dmic_ts_local_tsctrl_addr(void)
+{
+	return TIMESTAMP_BASE + TS_DMIC_LOCAL_TSCTRL;
+}
+
+static inline uint32_t dmic_ts_local_offs_addr(void)
+{
+	return TIMESTAMP_BASE + TS_DMIC_LOCAL_OFFS;
+}
+
+static inline uint32_t dmic_ts_local_sample_addr(void)
+{
+	return TIMESTAMP_BASE + TS_DMIC_LOCAL_SAMPLE;
+}
+
+static inline uint32_t dmic_ts_local_walclk_addr(void)
+{
+	return TIMESTAMP_BASE + TS_DMIC_LOCAL_WALCLK;
+}
+
+static inline uint32_t dmic_ts_tscc_addr(void)
+{
+	return TIMESTAMP_BASE + TS_DMIC_TSCC;
+}
+
+static int dmic_ts_config(struct dai *dai, struct sof_timestamp_cfg *cfg)
+{
+	if (cfg->type != SOF_DAI_INTEL_DMIC) {
+		trace_dmic_error("dmic_ts_config(): Illegal DAI type");
+		return -EINVAL;
+	}
+
+	cfg->walclk_rate = DMIC_HW_IOCLK;
+
+	return 0;
+}
+
+static int dmic_ts_start(struct dai *dai, struct sof_timestamp_cfg *cfg)
+{
+	uint32_t cdmas;
+	uint32_t addr = dmic_ts_local_tsctrl_addr();
+
+	/* Set DMIC timestamp registers */
+	spin_lock(dai->lock);
+
+	/* First point CDMAS to GPDMA channel that is used by DMIC */
+	cdmas = TS_LOCAL_TSCTRL_CDMAS(cfg->dma_chan_index +
+		cfg->dma_chan_count * cfg->dma_id);
+	io_reg_write(addr, cdmas);
+
+	/* Request on demand timestamp */
+	io_reg_write(addr, TS_LOCAL_TSCTRL_ODTS_BIT | cdmas);
+
+	spin_unlock(dai->lock);
+	return 0;
+}
+
+static int dmic_ts_get(struct dai *dai, struct sof_timestamp_cfg *cfg,
+		       struct sof_timestamp_data *tsd)
+{
+	/* Read DMIC timestamp registers */
+	uint32_t ntk;
+	uint32_t addr;
+	uint32_t tsctrl = dmic_ts_local_tsctrl_addr();
+
+	/* Read SSP timestamp registers */
+	spin_lock(dai->lock);
+	ntk = io_reg_read(tsctrl) & TS_LOCAL_TSCTRL_NTK_BIT;
+	if (ntk) {
```

Comment:
better to fail early here and check for if (!ntk). This will also help you reducing one level of indentation.

Reply:
I did it this way to have a single spin_lock(); do_things(); spin_unlock(); flow. But OTOH I'm not sure using spin lock is needed here. I'm not sure that another DAI instance would mess things up if it could get in between these I/O operations. I'm a signal processing guy and not in my comfort zone with drivers. So I'd not change this but will do if the experts here confirm it's the right thing to do. OK?

Reply:
@singalsu @dbaluta means something like this
```
spin_lock()
ntk = io_read(X);
if (!ntk)
    goto out;

/* set up clock since ntk is valid */

out:
spin_unlock();
```

Reply:
OK I got it now. I guess I introduced "addr" to avoid long lines and now with less indent it's no more needed...

This change is in the latest pushed version.